### ISSUE TITLE
Enqueue from .NET Core, execute on classic .NET

### ISSUE THREAD

**Issue by pieceofsummer** [HangfireIO/Hangfire#816]
There's a cross-framework type name serialization issue. 

Common built-in types (like String, Int32 etc.), which were previously defined in `mscorlib`, are now sitting in `System.Private.CoreLib` on .NET Core, so the classic .NET framework will fail to deserialize them because of a missing assembly. The problem is actually one-sided, because .NET Core has type redirection and will successfully deserialize mscorlib types.

It sure can be workarounded by providing a custom `JsonConverter` for JobHelper, but it would be nice for this to be handled out-of-the-box.

For reference, the converter itself might be like this:
```c#
public class CompatibleTypeConverter : JsonConverter
{
    public override bool CanConvert(Type objectType) => typeof(Type).IsAssignableFrom(objectType);

    public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
    {
        switch (reader.TokenType)
        {
            case JsonToken.Null:
                return null;

            case JsonToken.String:
                return Type.GetType((string)reader.Value, true);

            default:
                throw new JsonSerializationException($"Unexpected token when reading type name: {reader.TokenType}");
        }
    }

    public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
    {
        if (value == null)
        {
            writer.WriteNull();
            return;
        }

        var type = (Type)value;
        var typeName = type.AssemblyQualifiedName;

#if NETSTANDARD1_3 || NETSTANDARD1_4 || NETSTANDARD1_5 || NETSTANDARD1_6 || NETCOREAPP1_0 || NETCOREAPP1_1 // || ...
        var assemblyName = type.GetTypeInfo().Assembly.GetName();
        if (assemblyName.Name == "System.Private.CoreLib")
        {
            // AssemblyName is copied, so we can overwrite Name in-place
            assemblyName.Name = "mscorlib";

            var mscorlibTypeName = Assembly.CreateQualifiedName(assemblyName.FullName, type.FullName);

            if (type == Type.GetType(mscorlibTypeName, false))
            {
                // type is redirected
                typeName = mscorlibTypeName;
            }
        }
#endif

        writer.WriteValue(typeName);
    }
}
```

**Comment by burningice2866** [comment]
.NET Standard sure has brought a lot of headache along the way :)

**Comment by mkonstan** [comment]
Hi,

Can you post a workaround sample. I am experiencing the same issue where using asp.net core on front end with net462 on backed.

Thanks.

**Comment by rfcdejong** [comment]
Seems this works only for the root, I solved it with another workaround, including a part from 
https://stackoverflow.com/questions/19666511/how-to-create-a-serializationbinder-for-the-binary-formatter-that-handles-the-mo/19927484#19927484

Which has also been implemented by Microsoft in Bing Maps
https://github.com/Microsoft/Bing-Maps-Fleet-Tracker/blob/master/Backend/src/Trackable.Func/Shared/AdvancedSerializationBinder.cs

Mine is just a bit different, to support the Compatibility between NETCORE and NETFRAMEWORK.

CustomSerializationBinder with ISerializationBinder interface implemented. Then in the BindToType the following code

```
       public virtual Type BindToType(string assemblyName, string typeName)
        {
**#if NETFRAMEWORK
            if (assemblyName == "System.Private.CoreLib")
                assemblyName = "mscorlib";

            typeName = typeName.Replace("System.Private.CoreLib", "mscorlib");
#elif NETCOREAPP
            if (assemblyName == "mscorlib")
                assemblyName = "System.Private.CoreLib";

            typeName = typeName.Replace("mscorlib", "System.Private.CoreLib");
#endif**

            var m = Regex.Match(typeName, @"^(?<gen>[^\[]+)\[\[(?<type>[^\]]*)\](,\[(?<type>[^\]]*)\])*\]$");
            if (m.Success)
            { // generic type
                var gen = GetFlatTypeMapping(assemblyName, m.Groups["gen"].Value);
                var genArgs = m.Groups["type"]
                    .Captures
#if NETFRAMEWORK
                    .OfType<Capture>()
#endif
                    .Select(c =>
                    {
                        var m2 = Regex.Match(c.Value, @"^(?<type>(\w+(\.?\+?\w+)+))\s*,\s*(?<assembly>[\w\.]+)?$");
                        return m2.Success ? BindToType(m2.Groups["assembly"].Value, m2.Groups["type"].Value) : null;
                    })
                    .ToArray();
                return gen.MakeGenericType(genArgs);
            }
            return GetFlatTypeMapping(assemblyName, typeName);
        }

Construct the CustomSerializationBinder and assign it in JsonSerializerSettings.
```

**Comment by rfcdejong** [comment]
If that is the case, you should close the issue. My reply was also for using the JsonSerializer.Convert for everything, also outside Hangfire :)

**Comment by odinserj** [comment]
This issue is resolved by the https://github.com/HangfireIO/Hangfire/commit/904c53549e6ed19b6af5f908b82089025e666848 commit and released with version [1.6.22](https://github.com/HangfireIO/Hangfire/releases/tag/v1.6.22).
